components/UserObserver.php

```diff
@@ -17,112 +17,239 @@
+				\Yii::log("Unable to update user. Error: ". $e->getMessage(), \CLogger::LEVEL_ERROR);
+				throw new Exception("Unable to save user contact: ".print_r($user->getErrors(),true));
 			}
 		}
 	}
 
-	public function is_stale($username)
+
+	/**
+     * Finds User by username
+     *
+     * @param string $username
+     * @return User
+     */
+	private function getUser(string $username, int $institution_authentication_id): User
+	{
+		$criteria = new \CDbCriteria();
+		$criteria->join = 'JOIN user_authentication ua ON t.id = ua.user_id';
+		$criteria->addCondition('ua.username = :username');
+		$criteria->params[':username'] = $username;
+		$criteria->addCondition('ua.institution_authentication_id = :institution_authentication_id');
+		$criteria->params[':institution_authentication_id'] = $institution_authentication_id;
+		$user = \User::model()->find($criteria);
+		return $user;
+	}
+
+	/**
+     * Finds UserAuthentication by username
+     *
+     * @param string $username
+     * @return UserAuthentication
+     */
+	private function getUserAuthentication(string $username, int $institution_authentication_id): UserAuthentication
+	{
+		$criteria = new \CDbCriteria();
+		$criteria->join = 'JOIN institution_authentication ia ON t.institution_authentication_id = ia.id';
+		$criteria->addCondition('username = :username');
+		$criteria->params[':username'] = $username;
+		$criteria->addCondition('institution_authentication_id = :institution_authentication_id');
+		$criteria->params[':institution_authentication_id'] = $institution_authentication_id;
+		$user_authentication = \UserAuthentication::model()->find($criteria);
+		return $user_authentication;
+	}
+
+	/**
+     * Saves new User data which is coming from $remote_user
+     *
+     * @param User $user
+	 * @param UserAuthentication $user_authentication
+	 * @param array $remote_user
+     * @return User
+     */
+	private function saveUser(User $user, UserAuthentication $user_authentication, array $remote_user): User
+	{
+		$user->code = $remote_user['code'];
+		$user_authentication->username = $remote_user['username'];
+		$user->first_name = $remote_user['first_name'];
+		$user->last_name = $remote_user['last_name'];
+		$mehstaffdb_default_email = Yii::app()->params['mehstaffdb_default_email'];
+		if(strlen($mehstaffdb_default_email) != 0) {
+			$user->email = $mehstaffdb_default_email;
+		}
+		$user->title = $remote_user['title'];
+		$user->qualifications = $remote_user['qualifications'];
+		$user->role = $remote_user['role'];
+		$user->doctor_grade_id = $this->getDoctorGradeFromJobTitle($remote_user['role']);
+		if(isset($remote_user['registration_code']) && isset($remote_user['registration_code'][0]['ProfessionalRegistration'])) {
+			$user->registration_code = $this->getGMCRegistrationNumber($remote_user['registration_code'][0]['ProfessionalRegistration']);
+		}
+		//$user->password = 'faed6633f5a86241f3e0c2bb2bb768fd';
```

Comment:
Passwords should never be saved in plaintext in a public repo!!!

also commented out code should be removed if there is no reason for it. If there is a reason please add that reason as another comment

Reply:
Removed commented password